### DISCUSSION TITLE
Monoidal products

### DISCUSSION THREAD

**Discussion by suhr** [jameshaydon/lawvere#13] [Ideas]
In

```
ar squarePlusOne : Int --> Int = identity * identity + 1

ar squarePlusOne' : Int --> Int = * + 1
```

there's an implicit duplication of the input value.  Did you consider making it explicit by introducing a strict monoidal product and functions `dup` and `drop`?

So this example would be something like

```
ar squarePlusOne : Int --> Int = dup (id),(+ 1) (*)
```

or with infix sugar

```
ar squarePlusOne : Int --> Int = dup id * (+ 1)
```

This allows working with affine or linear types (you just don't allow using `dup` or `drop` for them). Also, it makes the language closer to a concatenative language (there's an example what a concatenative language can feel like: https://suhr.github.io/wcpl/intro.html).

**Comment by jameshaydon** [comment]
I think this is a great idea; I would like to support more general monoidal products.
